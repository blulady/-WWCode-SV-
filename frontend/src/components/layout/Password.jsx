import React, { useState, useEffect } from "react";
import cancelIcon from "../../images/cancel_24px.png";
import checkIcon from "../../images/check_circle_24px.png";
import styles from "./Password.module.css";

const Password = ({ setPwd , pwdLabel= "Password *"}) => {
  const [password, setPassword] = useState("");
  const [hidden, setHidden] = useState(true);
  const [pwdValidation, setPwdValidation] = useState(false);
  const passwordValidationDescriptions = [
    "8-50 characters",
    "At least  1 upper case letter",
    "At least 1 lower case letter",
    "At least 1 numeric character",
  ];
  const [passwdValidStatus, setPasswdValidStatus] = useState([
    false,
    false,
    false,
    false,
  ]);

  useEffect(() => {
    checkValidation();
  }, [password]);

  // function for validating password
  const validatePassword = (e) => {
    const regexs = [
      RegExp(/^(?=.{8,50}$)/),
      RegExp(/(?=.*[A-Z])/),
      RegExp(/(?=.*[a-z])/),
      RegExp(/(?=.*[0-9])/),
    ];
    const validStatus = regexs.map((regex) => regex.test(e.target.value));
    setPasswdValidStatus(() => validStatus);
    setPassword(e.target.value);
  };
  // function to handle show/hide toggle
  const handleShow = (e) => {
    setHidden(!hidden);
  };

  const checkValidation = () => {
    if (passwdValidStatus.every((val) => val === true)) {
      setPwd(password);
    } else {
      setPwd("");
    }
  };

  return (
    <div className='form-group' onFocus={() => setPwdValidation(true)}>
      <label className={`pb-1 ${styles.label}`} htmlFor='Password'>{pwdLabel}</label>
      <div>
        <div className='input-group'>
          <input
            type={hidden ? "password" : "text"}
            name='password'
            style={{ borderRight: "none" }}
            className='form-control'
            data-testId="password"
            id='password'
            value={password}
            autoComplete="password"
            onChange={validatePassword}
          />
          <div className='input-group-append' style={{ borderLeft: "none" }}>
            <span className='show-hide input-group-text' onClick={handleShow}>
              <u>{hidden ? "SHOW" : "HIDE"}</u>
            </span>
          </div>
        </div>
      </div>
      <div className={styles.validations}>
        <aside style={{ display: pwdValidation ? "inline" : "none" }}>
          {passwdValidStatus.map((x, idx) => {
            const iconSrc = x ? checkIcon : cancelIcon;
            return (
              <div key={idx}>
                <img
                  src={iconSrc}
                  alt='Invalid Password Icon'
                  style={{ width: "17.5px", height: "17.5px" }}
                />
                <span className={styles.pwddesc}>
                  {passwordValidationDescriptions[idx]}
                </span>
              </div>
            );
          })}
        </aside>
      </div>
    </div>
  );
};

export default Password;
