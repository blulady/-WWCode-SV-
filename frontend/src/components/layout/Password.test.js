import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import Password from './Password';

const mockSetPassword = jest.fn();

describe('Password', () => {
    beforeEach(() => {
        mockSetPassword.mockClear(); 
    });

    test('it renders without crashing', () => {
        const { container } = render(<Password setPwd={mockSetPassword} />);

        expect(container).toMatchSnapshot();
    });

    test.skip('it validates password on change', () => {
        const { container, getByTestId } = render(<Password setPwd={mockSetPassword} />);
        const inputPwd = getByTestId('password');
        
        fireEvent.change(inputPwd, { target: { value: 'Password123' } });

        // Why is the invalid icon showing up in the Password component?
        //const invalidIcon = container.querySelector('.validation-img');
        //expect(invalidIcon).not.toBeInTheDocument();
    });

    test('it shows password on show button click', () => {
        const { container, getByTestId } = render(<Password setPwd={mockSetPassword} />);
        const inputPwd = getByTestId('password');
        const showHide = getByTestId('show-hide');

        fireEvent.click(showHide);

        expect(inputPwd.type).toBe('text');
    });
});