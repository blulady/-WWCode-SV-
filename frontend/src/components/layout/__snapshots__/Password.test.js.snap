// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password it renders without crashing 1`] = `
<div>
  <div
    class="form-group"
  >
    <label
      class="pb-1 label"
      for="Password"
    >
      Password *
    </label>
    <div
      class="row"
    >
      <div
        class="col"
      >
         
        <div
          class="input-group pwd-input-group"
        >
          <input
            autocomplete="password"
            class="form-control input-pwd"
            data-testid="password"
            id="password"
            name="password"
            type="password"
            value=""
          />
          <div
            class="input-group-append"
          >
            <span
              class="input-group-text show-hide"
            >
              <u
                data-testid="show-hide"
              >
                SHOW
              </u>
            </span>
          </div>
        </div>
      </div>
      <div
        class="display-none"
      >
         
        <aside>
          <div>
            <img
              alt="Invalid Password Icon"
              class="validation-img"
              src="cancel_24px.png"
            />
            <span
              class="pwddesc"
            >
              8-50 characters
            </span>
          </div>
          <div>
            <img
              alt="Invalid Password Icon"
              class="validation-img"
              src="cancel_24px.png"
            />
            <span
              class="pwddesc"
            >
              At least  1 upper case letter
            </span>
          </div>
          <div>
            <img
              alt="Invalid Password Icon"
              class="validation-img"
              src="cancel_24px.png"
            />
            <span
              class="pwddesc"
            >
              At least 1 lower case letter
            </span>
          </div>
          <div>
            <img
              alt="Invalid Password Icon"
              class="validation-img"
              src="cancel_24px.png"
            />
            <span
              class="pwddesc"
            >
              At least 1 numeric character
            </span>
          </div>
        </aside>
      </div>
    </div>
     
  </div>
</div>
`;
